x-pack/auditbeat/module/system/login/login.go

```diff
@@ -0,0 +1,193 @@
+
+	config := defaultConfig
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, errors.Wrapf(err, "failed to unpack the %v/%v config", moduleName, metricsetName)
+	}
+
+	bucket, err := datastore.OpenBucket(bucketName)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open persistent datastore")
+	}
+
+	ms := &MetricSet{
+		BaseMetricSet: base,
+		config:        config,
+		log:           logp.NewLogger(metricsetName),
+	}
+
+	ms.utmpReader, err = NewUtmpFileReader(ms.log, bucket, config.UtmpFilePattern)
+	if err != nil {
+		return nil, err
+	}
+
+	return ms, nil
+}
+
+// Close cleans up the MetricSet when it finishes.
+func (ms *MetricSet) Close() error {
+	return ms.utmpReader.Close()
+}
+
+// Fetch collects any new login records from /var/log/wtmp. It is invoked periodically.
+func (ms *MetricSet) Fetch(report mb.ReporterV2) {
+	loginRecords, err := ms.utmpReader.ReadNew()
+	if err != nil {
+		ms.log.Error(err)
+		report.Error(err)
+		return
+	}
+	ms.log.Debugf("%d new login records.", len(loginRecords))
+
+	for _, loginRecord := range loginRecords {
+		report.Event(ms.loginEvent(loginRecord))
+	}
+
+	// Save new state to disk
+	if len(loginRecords) > 0 {
+		err := ms.utmpReader.saveStateToDisk()
+		if err != nil {
+			ms.log.Error(err)
+			report.Error(err)
+			return
+		}
+	}
+}
+
+func (ms *MetricSet) loginEvent(loginRecord LoginRecord) mb.Event {
+	event := mb.Event{
+		Timestamp: loginRecord.Timestamp,
+		RootFields: common.MapStr{
+			"event": common.MapStr{
+				"type":   eventTypeEvent,
+				"action": loginRecord.Type,
+				"origin": loginRecord.Origin,
+			},
+		},
+		MetricSetFields: loginRecord.toMapStr(),
+	}
+
+	if loginRecord.Username != "" {
+		event.RootFields.Put("user.name", loginRecord.Username)
+
+		if loginRecord.UID != -1 {
+			event.RootFields.Put("user.id", loginRecord.UID)
+		}
+	}
+
+	if loginRecord.PID != -1 {
+		event.RootFields.Put("process.pid", loginRecord.PID)
+	}
+
+	if loginRecord.IP != nil {
+		event.RootFields.Put("source.ip", loginRecord.IP)
+	}
+
+	if loginRecord.Hostname != "" && loginRecord.Hostname != loginRecord.IP.String() {
+		event.RootFields.Put("source.hostname", loginRecord.Hostname)
+	}
+
+	var eventSummary string
+
+	switch loginRecord.Type {
+	case LoginRecordTypeBoot:
+		eventSummary = "System boot"
+	case LoginRecordTypeShutdown:
+		eventSummary = "System shutdown"
+	case LoginRecordTypeUserLogin:
+		eventSummary = fmt.Sprintf("Login by user %v (UID: %d) on %v (PID: %d) from %v (IP: %v).",
+			loginRecord.Username, loginRecord.UID, loginRecord.TTY, loginRecord.PID,
+			loginRecord.Hostname, loginRecord.IP)
+	case LoginRecordTypeUserLogout:
+		eventSummary = fmt.Sprintf("Logout by user %v (UID: %d) on %v (PID: %d) from %v (IP: %v).",
+			loginRecord.Username, loginRecord.UID, loginRecord.TTY, loginRecord.PID,
+			loginRecord.Hostname, loginRecord.IP)
+	}
+
+	event.RootFields.Put("event.summary", eventSummary)
```

Comment:
I think we're going with `message` instead of `event.summary` for this. @webmat @MikePaquette, correct? Or do you prefer to do the rename in a follow up PR for the whole module?

Reply:
Since this is the first time we're introducing this field we can settle on a name in this PR. I personally would want to keep it as `summary`, because it's different from `message` and there is a place for both of them in the same document. In the PR description above I'm arguing the details (pasting here):

event.summary contains a human-readable summary of the event, e.g. Login by user vagrant (UID: 1000) on pts/0 (PID: 9355) from 10.0.2.2 (IP: 10.0.2.2).. It doesn't add any information that is not already in any of the other fields, it's there to make it easy to consume this data together with data from the other system module metricsets e.g. in a table in Kibana. I chose not to use the message base field from ECS, since in my mind a message is something "real" that comes from outside the Beat e.g. a log message. summary is not original/authoritative and doesn't contain any new information. It can be removed without losing anything. It could be used to summarize a message field as well, e.g. the unwieldy access log messages from web servers.

Reply:
The advantage of `message` is that it would work OOB in the log viewer and similar UIs. We could maybe change the log viewer to also look for this field. Something to discuss.

Reply:
As discussed, I've changed to using `message` instead of `event.summary` since that's the ECS standard.